usr/lib/xtable/libxtable.c

```diff
@@ -0,0 +1,216 @@
+}
+
+#define LIBXTABLE_INITER_CB(hook)				   	\
+static void libxtable_##hook##_init(struct xt_entry_##hook *hook)   	\
+{								   	\
+	libxtable_run(L, "init", libxtable_##hook##_key, 1, 0);		\
+}
+
+#define LIBXTABLE_PARSER_CB(hook)			   		\
+static int libxtable_##hook##_parse(int c, char **argv, int invert, unsigned int *flags,	\
+					const void *entry, struct xt_entry_##hook **hook)	\
+{							   		\
+	return libxtable_run(L, "parse", libxtable_##hook##_key, 0, 1);	\
+}
+
+#define LIBXTABLE__FINALCHECKER_CB(hook)				\
+static void libxtable_##hook##_finalcheck(unsigned int flags)		\
+{							  		\
+	libxtable_run(L, "final_check", libxtable_##hook##_key, 0, 0);	\
+}
+
+#define LIBXTABLE_PRINTER_CB(hook)  					\
+static void libxtable_##hook##_print(const void *entry, const struct xt_entry_##hook *hook, int numeric)	\
+{							   		\
+	libxtable_run(L, "print", libxtable_##hook##_key, 0, 0);	\
+}
+
+#define LIBXTABLE_SAVER_CB(hook)				\
+static void libxtable_##hook##_save(const void *entry, const struct xt_entry_##hook *hook)  \
+{								\
+	libxtable_run(L, "save", libxtable_##hook##_key, 0, 0);	\
+}
+
+#define LIBXTABLE_NEWCB(hook)  		\
+	LIBXTABLE_HELPER_CB(hook)   	\
+	LIBXTABLE_INITER_CB(hook)   	\
+	LIBXTABLE_PARSER_CB(hook)   	\
+	LIBXTABLE__FINALCHECKER_CB(hook)	\
+	LIBXTABLE_PRINTER_CB(hook)  	\
+	LIBXTABLE_SAVER_CB(hook)
+
+LIBXTABLE_NEWCB(match);
+LIBXTABLE_NEWCB(target);
+
+
+#define LIBXTABLE_NEWREG(hook)					\
+static struct xtables_##hook libxtable_##hook##_reg = {		\
+	.version = XTABLES_VERSION,				\
+	.name = LUAXTABLE_MODULE,					\
+	.size = XT_ALIGN(sizeof(luaxtable_info_t)),		\
+	.userspacesize = 0,					\
+	.help = libxtable_##hook##_help,			\
+	.init = libxtable_##hook##_init,			\
+	.parse = libxtable_##hook##_parse,		 	\
+	.final_check = libxtable_##hook##_finalcheck,	  	\
+	.print = libxtable_##hook##_print,		  	\
+	.save = libxtable_##hook##_save			 	\
+}
+
+LIBXTABLE_NEWREG(match);
+LIBXTABLE_NEWREG(target);
+
+static inline void libxtable_setregistry(lua_State *L, int idx, const char *key)
+{
+	lua_pushvalue(L, idx);
+	lua_rawsetp(L, LUA_REGISTRYINDEX, key);
+}
+
+static inline void libxtable_putregistry(lua_State *L, const char *key)
+{
+	lua_pushnil(L);
+	lua_rawsetp(L, LUA_REGISTRYINDEX, key);
+}
+
+static inline int libxtable_checkint(lua_State *L, int idx, const char *key)
+{
+	if(lua_getfield(L, idx, key) != LUA_TNUMBER) {
+		fprintf(stderr, "invalid \'%s\' in match\n", key);
+		return 0;
+	}
+	int ret = lua_tointeger(L, -1);
+	lua_pop(L, 1);
+	return ret;
+}
+
+#define LIBXTABLE_LIB_CB(hook)				\
+static int libxtable_##hook##_lib(lua_State *L)		\
+{							\
+	luaL_checktype(L, 1, LUA_TTABLE);		\
+							\
+	libxtable_##hook##_reg.revision = libxtable_checkint(L, 1, "revision");		\
+	libxtable_##hook##_reg.family = libxtable_checkint(L, 1, "family");		\
+								\
+	xtables_register_##hook(&libxtable_##hook##_reg);	\
+	libxtable_setregistry(L, 1, libxtable_##hook##_key);	\
```

Comment:
IMHO, we don't need this function as well.. it's just two lines used only here and clean enough..